File: athlon-user/src/lib/api/academyPost.ts
import { api, fetchClient } from './client';

export interface AcademyPost {
  postId?: number;
  postUuid: string;
  organizationId?: number;
  organizationUuid: string;
  title: string;
  content: string;
  postType: 'BLOG' | 'VIDEO' | 'GALLERY' | 'ANNOUNCEMENT';
  mediaUrls?: string;
  coverImageUrl?: string;
  youtubeVideoUrl?: string;
  youtubeVideoId?: string;
  targetScope: 'ALL' | 'CENTRE' | 'BATCH';
  centreUuid?: string;
  centreName?: string;
  batchUuid?: string;
  batchName?: string;
  category?: string;
  tags?: string;
  authorUserId?: number;
  authorUserUuid?: string;
  authorName: string;
  authorRole: 'ADMIN' | 'COACH' | 'STUDENT';
  authorAvatar?: string;
  isPinned?: boolean;
  likesCount: number;
  commentsCount: number;
  viewsCount: number;
  isLikedByCurrentUser?: boolean;
  status: 'PUBLISHED' | 'DRAFT' | 'ARCHIVED';
  createdAt: string;
  updatedAt?: string;
}

export interface AcademyPostComment {
  commentId?: number;
  commentUuid: string;
  postId?: number;
  postUuid: string;
  organizationUuid: string;
  authorUserId?: number;
  authorUserUuid: string;
  authorName: string;
  authorRole: string;
  authorAvatar?: string;
  commentText: string;
  createdAt: string;
}

export interface CreateAcademyPostPayload {
  organizationUuid: string;
  title: string;
  content: string;
  postType: 'BLOG' | 'VIDEO' | 'GALLERY' | 'ANNOUNCEMENT';
  mediaUrls?: string;
  coverImageUrl?: string;
  youtubeVideoUrl?: string;
  targetScope?: 'ALL' | 'CENTRE' | 'BATCH';
  centreUuid?: string;
  centreName?: string;
  batchUuid?: string;
  batchName?: string;
  category?: string;
  tags?: string;
  authorName?: string;
  authorRole?: string;
  authorAvatar?: string;
}

export interface UpdateAcademyPostPayload {
  postUuid: string;
  title?: string;
  content?: string;
  postType?: string;
  mediaUrls?: string;
  coverImageUrl?: string;
  youtubeVideoUrl?: string;
  targetScope?: string;
  centreUuid?: string;
  centreName?: string;
  batchUuid?: string;
  batchName?: string;
  category?: string;
  tags?: string;
  isPinned?: boolean;
  status?: string;
}

export interface AddPostCommentPayload {
  postUuid: string;
  commentText: string;
  authorName?: string;
  authorRole?: string;
  authorAvatar?: string;
}

export const AcademyPostService = {
  getPosts: async (
    orgUuid: string,
    type?: string,
    scope?: string,
    batchUuid?: string,
    centreUuid?: string,
    search?: string
  ): Promise<AcademyPost[]> => {
    let url = `/api/identity/academy/posts/org/${orgUuid}?`;
    if (type && type !== 'ALL') url += `type=${type}&`;
    if (scope && scope !== 'ALL') url += `scope=${scope}&`;
    if (batchUuid && batchUuid !== 'ALL') url += `batchUuid=${batchUuid}&`;
    if (centreUuid && centreUuid !== 'ALL') url += `centreUuid=${centreUuid}&`;
    if (search && search.trim()) url += `search=${encodeURIComponent(search.trim())}&`;

    const res = await api.get<any>(url);
    if (Array.isArray(res)) return res;
    if (res && Array.isArray(res.data)) return res.data;
    return [];
  },

  getPostByUuid: async (postUuid: string): Promise<AcademyPost | null> => {
    const res = await api.get<any>(`/api/identity/academy/posts/${postUuid}`);
    return res?.data || res || null;
  },

  createPost: async (payload: CreateAcademyPostPayload): Promise<AcademyPost> => {
    const res = await api.post<any>('/api/identity/academy/posts/create', payload);
    return res?.data || res;
  },

  createPostMultipart: async (
    payload: CreateAcademyPostPayload,
    coverFile?: File | null,
    galleryFiles?: File[]
  ): Promise<AcademyPost> => {
    const formData = new FormData();
    if (payload.organizationUuid) formData.append('organizationUuid', payload.organizationUuid);
    if (payload.title) formData.append('title', payload.title);
    if (payload.content) formData.append('content', payload.content);
    if (payload.postType) formData.append('postType', payload.postType);

    if (payload.targetScope) formData.append('targetScope', payload.targetScope);
    if (payload.centreUuid && payload.centreUuid.trim()) formData.append('centreUuid', payload.centreUuid.trim());
    if (payload.centreName && payload.centreName.trim()) formData.append('centreName', payload.centreName.trim());
    if (payload.batchUuid && payload.batchUuid.trim()) formData.append('batchUuid', payload.batchUuid.trim());
    if (payload.batchName && payload.batchName.trim()) formData.append('batchName', payload.batchName.trim());
    if (payload.category && payload.category.trim()) formData.append('category', payload.category.trim());
    if (payload.tags && payload.tags.trim()) formData.append('tags', payload.tags.trim());
    if (payload.youtubeVideoUrl && payload.youtubeVideoUrl.trim()) formData.append('youtubeVideoUrl', payload.youtubeVideoUrl.trim());
    if (payload.coverImageUrl && payload.coverImageUrl.trim()) formData.append('coverImageUrl', payload.coverImageUrl.trim());
    if (payload.mediaUrls && payload.mediaUrls.trim()) formData.append('mediaUrls', payload.mediaUrls.trim());
    if (payload.authorName && payload.authorName.trim()) formData.append('authorName', payload.authorName.trim());
    if (payload.authorRole && payload.authorRole.trim()) formData.append('authorRole', payload.authorRole.trim());

    if (coverFile) {
      formData.append('coverFile', coverFile);
    }

    if (galleryFiles && galleryFiles.length > 0) {
      galleryFiles.forEach((gf) => {
        formData.append('galleryFiles', gf);
      });
    }

    const res = await fetchClient<any>('/api/identity/academy/posts/createMultipart', {
      method: 'POST',
      body: formData,
    });
    return res?.data || res;
  },

  uploadMediaFiles: async (files: File[]): Promise<string[]> => {
    const formData = new FormData();
    files.forEach((f) => formData.append('files', f));
    const res = await fetchClient<any>('/api/identity/academy/posts/uploadMedia', {
      method: 'POST',
      body: formData,
    });
    return res?.data || res || [];
  },

  updatePost: async (payload: UpdateAcademyPostPayload): Promise<AcademyPost> => {
    const res = await api.post<any>('/api/identity/academy/posts/update', payload);
    return res?.data || res;
  },

  deletePost: async (postUuid: string): Promise<void> => {
    await api.post<void>(`/api/identity/academy/posts/delete/${postUuid}`, {});
  },

  toggleLike: async (postUuid: string): Promise<boolean> => {
    const res = await api.post<any>(`/api/identity/academy/posts/${postUuid}/like`, {});
    return Boolean(res?.data ?? res);
  },

  addComment: async (payload: AddPostCommentPayload): Promise<AcademyPostComment> => {
    const res = await api.post<any>(`/api/identity/academy/posts/${payload.postUuid}/comment`, payload);
    return res?.data || res;
  },

  getComments: async (postUuid: string): Promise<AcademyPostComment[]> => {
    const res = await api.get<any>(`/api/identity/academy/posts/${postUuid}/comments`);
    if (Array.isArray(res)) return res;
    if (res && Array.isArray(res.data)) return res.data;
    return [];
  },

  togglePin: async (postUuid: string): Promise<AcademyPost> => {
    const res = await api.post<any>(`/api/identity/academy/posts/${postUuid}/pin`, {});
    return res?.data || res;
  },
};

File: ATHLON/IDENTITYSERVICE/src/main/java/com/athlon/identityservice/organization/entity/AcademyPost.java
package com.athlon.identityservice.organization.entity;

import java.time.LocalDateTime;
import java.util.UUID;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;

@Entity
@Table(name = "academy_posts")
public class AcademyPost {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "post_id", updatable = false, nullable = false)
    private Long postId;

    @Column(name = "post_uuid", updatable = false, nullable = false, unique = true)
    private UUID postUuid;

    @Column(name = "organization_id", nullable = false)
    private Long organizationId;

    @Column(name = "organization_uuid", nullable = false)
    private UUID organizationUuid;

    @Column(name = "title", nullable = false, length = 255)
    private String title;

    @Column(name = "content", nullable = false, columnDefinition = "TEXT")
    private String content;

    @Column(name = "post_type", nullable = false, length = 30)
    private String postType; // BLOG, VIDEO, GALLERY, ANNOUNCEMENT

    @Column(name = "media_urls", columnDefinition = "TEXT")
    private String mediaUrls; // Comma-separated or JSON array of image URLs

    @Column(name = "cover_image_url", length = 500)
    private String coverImageUrl;

    @Column(name = "youtube_video_url", length = 500)
    private String youtubeVideoUrl;

    @Column(name = "youtube_video_id", length = 50)
    private String youtubeVideoId;

    @Column(name = "target_scope", nullable = false, length = 20)
    private String targetScope = "ALL"; // ALL, CENTRE, BATCH

    @Column(name = "centre_uuid")
    private UUID centreUuid;

    @Column(name = "centre_name", length = 150)
    private String centreName;

    @Column(name = "batch_uuid")
    private UUID batchUuid;

    @Column(name = "batch_name", length = 150)
    private String batchName;

    @Column(name = "category", length = 60)
    private String category; // Drills & Training, Match Analysis, Nutrition, Events & Tournaments, General

    @Column(name = "tags", length = 255)
    private String tags;

    @Column(name = "author_user_id")
    private Long authorUserId;

    @Column(name = "author_user_uuid")
    private UUID authorUserUuid;

    @Column(name = "author_name", nullable = false, length = 150)
    private String authorName;

    @Column(name = "author_role", nullable = false, length = 50)
    private String authorRole; // ADMIN, COACH, STUDENT

    @Column(name = "author_avatar", length = 500)
    private String authorAvatar;

    @Column(name = "is_pinned", nullable = false)
    private Boolean isPinned = false;

    @Column(name = "likes_count", nullable = false)
    private Integer likesCount = 0;

    @Column(name = "comments_count", nullable = false)
    private Integer commentsCount = 0;

    @Column(name = "views_count", nullable = false)
    private Integer viewsCount = 0;

    @Column(name = "status", nullable = false, length = 20)
    private String status = "PUBLISHED"; // PUBLISHED, DRAFT, ARCHIVED

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public AcademyPost() {
    }

    @PrePersist
    protected void onCreate() {
        if (this.postUuid == null) {
            this.postUuid = UUID.randomUUID();
        }
        if (this.isPinned == null) {
            this.isPinned = false;
        }
        if (this.likesCount == null) {
            this.likesCount = 0;
        }
        if (this.commentsCount == null) {
            this.commentsCount = 0;
        }
        if (this.viewsCount == null) {
            this.viewsCount = 0;
        }
        if (this.status == null) {
            this.status = "PUBLISHED";
        }
        if (this.targetScope == null) {
            this.targetScope = "ALL";
        }
        if (this.createdAt == null) {
            this.createdAt = LocalDateTime.now();
        }
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    public Long getPostId() {
        return postId;
    }

    public void setPostId(Long postId) {
        this.postId = postId;
    }

    public UUID getPostUuid() {
        return postUuid;
    }

    public void setPostUuid(UUID postUuid) {
        this.postUuid = postUuid;
    }

    public Long getOrganizationId() {
        return organizationId;
    }

    public void setOrganizationId(Long organizationId) {
        this.organizationId = organizationId;
    }

    public UUID getOrganizationUuid() {
        return organizationUuid;
    }

    public void setOrganizationUuid(UUID organizationUuid) {
        this.organizationUuid = organizationUuid;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public String getPostType() {
        return postType;
    }

    public void setPostType(String postType) {
        this.postType = postType;
    }

    public String getMediaUrls() {
        return mediaUrls;
    }

    public void setMediaUrls(String mediaUrls) {
        this.mediaUrls = mediaUrls;
    }

    public String getCoverImageUrl() {
        return coverImageUrl;
    }

    public void setCoverImageUrl(String coverImageUrl) {
        this.coverImageUrl = coverImageUrl;
    }

    public String getYoutubeVideoUrl() {
        return youtubeVideoUrl;
    }

    public void setYoutubeVideoUrl(String youtubeVideoUrl) {
        this.youtubeVideoUrl = youtubeVideoUrl;
    }

    public String getYoutubeVideoId() {
        return youtubeVideoId;
    }

    public void setYoutubeVideoId(String youtubeVideoId) {
        this.youtubeVideoId = youtubeVideoId;
    }

    public String getTargetScope() {
        return targetScope;
    }

    public void setTargetScope(String targetScope) {
        this.targetScope = targetScope;
    }

    public UUID getCentreUuid() {
        return centreUuid;
    }

    public void setCentreUuid(UUID centreUuid) {
        this.centreUuid = centreUuid;
    }

    public String getCentreName() {
        return centreName;
    }

    public void setCentreName(String centreName) {
        this.centreName = centreName;
    }

    public UUID getBatchUuid() {
        return batchUuid;
    }

    public void setBatchUuid(UUID batchUuid) {
        this.batchUuid = batchUuid;
    }

    public String getBatchName() {
        return batchName;
    }

    public void setBatchName(String batchName) {
        this.batchName = batchName;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getTags() {
        return tags;
    }

    public void setTags(String tags) {
        this.tags = tags;
    }

    public Long getAuthorUserId() {
        return authorUserId;
    }

    public void setAuthorUserId(Long authorUserId) {
        this.authorUserId = authorUserId;
    }

    public UUID getAuthorUserUuid() {
        return authorUserUuid;
    }

    public void setAuthorUserUuid(UUID authorUserUuid) {
        this.authorUserUuid = authorUserUuid;
    }

    public String getAuthorName() {
        return authorName;
    }

    public void setAuthorName(String authorName) {
        this.authorName = authorName;
    }

    public String getAuthorRole() {
        return authorRole;
    }

    public void setAuthorRole(String authorRole) {
        this.authorRole = authorRole;
    }

    public String getAuthorAvatar() {
        return authorAvatar;
    }

    public void setAuthorAvatar(String authorAvatar) {
        this.authorAvatar = authorAvatar;
    }

    public Boolean getIsPinned() {
        return isPinned;
    }

    public void setIsPinned(Boolean isPinned) {
        this.isPinned = isPinned;
    }

    public Integer getLikesCount() {
        return likesCount;
    }

    public void setLikesCount(Integer likesCount) {
        this.likesCount = likesCount;
    }

    public Integer getCommentsCount() {
        return commentsCount;
    }

    public void setCommentsCount(Integer commentsCount) {
        this.commentsCount = commentsCount;
    }

    public Integer getViewsCount() {
        return viewsCount;
    }

    public void setViewsCount(Integer viewsCount) {
        this.viewsCount = viewsCount;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }
}
